runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,504 @@
+  local lo = 1
+  local hi = #tokens
+  while lo < hi do
+    local mid = math.floor((lo + hi) / 2)
+    if tokens[mid].line < line then
+      lo = mid + 1
+    else
+      hi = mid
+    end
+  end
+  return lo
+end
+
+---@private
+---@return string[]
+local function modifiers_from_number(x, modifiers_table)
+  local modifiers = {}
+  for i = 0, #modifiers_table - 1 do
+    local bit = get_bit(x, i)
+    if bit == 1 then
+      table.insert(modifiers, 1, modifiers_table[i + 1])
+    end
+  end
+
+  return modifiers
+end
+
+--- Converts a raw token list to a list of highlight ranges used by the on_win callback
+--
+---@private
+---@return STTokenRange[]
+local function tokens_to_ranges(client, bufnr, data)
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local ranges = {}
+
+  local line
+  local start_char = 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local function get_byte_pos(char_pos)
+      return util._get_line_byte_from_position(bufnr, {
+        line = line,
+        character = char_pos
+      }, client.offset_encoding)
+    end
+
+    local start_col = get_byte_pos(start_char)
+    local end_col = get_byte_pos(start_char + data[i + 2])
+
+    if token_type then
+      ranges[#ranges + 1] = {
+        line = line,
+        start_col = start_col,
+        end_col = end_col,
+        type = token_type,
+        modifiers = modifiers,
+        extmark_added = false,
+        offset_encoding = client.offset_encoding,
+      }
+    end
+  end
+
+  return ranges
+end
+
+---@private
+---@param client_id number
+---@param bufnr number
+---@return nil|STState
+---@return nil|STBufferState
+local function get_state(client_id, bufnr)
+  local state = state_by_client[client_id]
+  if not state then
+    return nil, nil
+  end
+  return state, state.buffers[bufnr]
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/full[/delta]`
+
+---This function will parse the semantic token responses and set up the cache
+---(current_result). It also performs document synchronization by checking the
+---version of the document associated with the resulting request_id and only
+---performing work if the response is not out-of-date.
+---
+---Delta edits are applied if necessary, and new highlight ranges are calculated
+---and stored in the buffer state.
+---
+---Finally, a redraw command is issued to force nvim to redraw the screen to
+---pick up changed highlight tokens.
+local function on_full_response(err, response, ctx, _)
```

Comment:
mark as `---@private`

Reply:
Yeah, was fixing gen_vimdoc.py as you were reviewing :)